<a href="https://colab.research.google.com/github/ebtezcan/dsc-phase-3-project/blob/master/index_OSEMN_colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## OVERVIEW OF OSEMiN

<img src='https://raw.githubusercontent.com/jirvingphd/fsds_100719_cohort_notes/master/images/OSEMN.png' width=800>

<center><a href="https://www.kdnuggets.com/2018/02/data-science-command-line-book-exploring-data.html"> 
    </a></center>


> <font size=2em>The Data Science Process we'll be using during this section--OSEMiN (pronounced "OH-sum", rhymes with "possum").  This is the most straightforward of the Data Science Processes discussed so far.  **Note that during this process, just like the others, the stages often blur together.***  It is completely acceptable (and ***often a best practice!) to float back and forth** between stages as you learn new things about your problem, dataset, requirements, etc.  
It's quite common to get to the modeling step and realize that you need to scrub your data a bit more or engineer a different feature and jump back to the "Scrub" stage, or go all the way back to the "Obtain" stage when you realize your current data isn't sufficient to solve this problem. 
As with any of these frameworks, *OSEMiN is meant to be treated as guidelines, not law. 
</font>


### OSEMN DETAILS

**OBTAIN**

- This step involves understanding stakeholder requirements, gathering information on the problem, and finally sourcing data that we think will be necessary for solving this problem. 

**SCRUB**

- During this stage, we'll focus on preprocessing our data.  Important steps such as identifying and removing null values, dealing with outliers, normalizing data, and feature engineering/feature selection are handled around this stage.  The line with this stage really blurs with the _Explore_ stage, as it is common to only realize that certain columns require cleaning or preprocessing as a result of the visualzations and explorations done during Step 3.  

- Note that although technically, categorical data should be one-hot encoded during this step, in practice, it's usually done after data exploration.  This is because it is much less time-consuming to visualize and explore a few columns containing categorical data than it is to explore many different dummy columns that have been one-hot encoded. 

**EXPLORE**

- This step focuses on getting to know the dataset you're working with. As mentioned above, this step tends to blend with the _Scrub_ step mentioned above.  During this step, you'll create visualizations to really get a feel for your dataset.  You'll focus on things such as understanding the distribution of different columns, checking for multicollinearity, and other tasks liek that.  If your project is a classification task, you may check the balance of the different classes in your dataset.  If your problem is a regression task, you may check that the dataset meets the assumptions necessary for a regression task.  

- At the end of this step, you should have a dataset ready for modeling that you've thoroughly explored and are extremely familiar with.  

**MODEL**

- This step, as with the last two frameworks, is also pretty self-explanatory. It consists of building and tuning models using all the tools you have in your data science toolbox.  In practice, this often means defining a threshold for success, selecting machine learning algorithms to test on the project, and tuning the ones that show promise to try and increase your results.  As with the other stages, it is both common and accepted to realize something, jump back to a previous stage like _Scrub_ or _Explore_, and make some changes to see how it affects the model.  

**iNTERPRET**

- During this step, you'll interpret the results of your model(s), and communicate results to stakeholders.  As with the other frameworks, communication is incredibily important! During this stage, you may come to realize that further investigation is needed, or more data.  That's totally fine--figure out what's needed, go get it, and start the process over! If your results are satisfactory to all stakeholders involved, you may also go from this stage right into productionizing your model and automating processes necessary to support it.  





## PROCESS CHECKLIST


> Keep in mind that it is normal to jump between the OSEMN phases and some of them will blend together, like SCRUB and EXPLORE.

1. **[OBTAIN](#OBTAIN)**
    - Import data, inspect, check for datatypes to convert and null values
    - Display header and info.
    - Drop any unneeded columns, if known (`df.drop(['col1','col2'],axis=1,inplace=True`)
    <br><br>


2. **[SCRUB](#SCRUB)**
    - Recast data types, identify outliers, check for multicollinearity, normalize data**
    - Check and cast data types
        - [ ] Check for #'s that are store as objects (`df.info()`,`df.describe()`)
            - when converting to #'s, look for odd values (like many 0's), or strings that can't be converted.
            - Decide how to deal weird/null values (`df.unique()`, `df.isna().sum()`)
            - `df.fillna(subset=['col_with_nulls'],'fill_value')`, `df.replace()`
        - [ ] Check for categorical variables stored as integers.
            - May be easier to tell when you make a scatter plotm or `pd.plotting.scatter_matrix()`
            
    - [ ] Check for missing values  (df.isna().sum())
        - Can drop rows or colums
        - For missing numeric data with median or bin/convert to categorical
        - For missing categorical data: make NaN own category OR replace with most common category
    - [ ] Check for multicollinearity
        - Use seaborn to make correlation matrix plot 
        - Good rule of thumb is anything over 0.75 corr is high, remove the variable that has the most correl with the largest # of variables
    - [ ] Normalize data (may want to do after some exploring)
        - Most popular is Z-scoring (but won't fix skew) 
        - Can log-transform to fix skewed data
    
    
3. **[EXPLORE](#EXPLORE)**
    - [ ] Check distributions, outliers, etc**
    - [ ] Check scales, ranges (df.describe())
    - [ ] Check histograms to get an idea of distributions (df.hist()) and data transformations to perform.
        - Can also do kernel density estimates
    - [ ] Use scatter plots to check for linearity and possible categorical variables (`df.plot("x","y")`)
        - categoricals will look like vertical lines
    - [ ] Use `pd.plotting.scatter_matrix(df)` to visualize possible relationships
    - [ ] Check for linearity.
   
   
4. **[MODEL](#MODEL)**

    - **Fit an initial model:** 
        - Run an initial model and get results

    - **Holdout validation / Train/test split**
        - use sklearn `train_test_split`
    
    
5. **[iNTERPRET](#iNTERPRET)**
    - **Assessing the model:**
        - Assess parameters (slope,intercept)
        - Check if the model explains the variation in the data (RMSE, F, R_square)
        - *Are the coeffs, slopes, intercepts in appropriate units?*
        - *Whats the impact of collinearity? Can we ignore?*
        <br><br>
    - **Revise the fitted model**
        - Multicollinearity is big issue for lin regression and cannot fully remove it
        - Use the predictive ability of model to test it (like R2 and RMSE)
        - Check for missed non-linearity
        
       
6. **Interpret final model and draw >=3 conclusions and recommendations from dataset**

<div style="display:block;border-bottom:solid red 3px;padding:1.4em;color:red;font-size:30pt;display:inline-block;line-height:1.5em;">
DELETE THIS CELL AND EVERYTHING ABOVE FROM YOUR FINAL NOTEBOOK
</div>

# Final Project Submission

Please fill out:
* Student name: 
* Student pace: self paced / part time / full time:
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:
* Video of 5-min Non-Technical Presentation:

## TABLE OF CONTENTS 

*Click to jump to matching Markdown Header.*<br><br>
 
- **[Introduction](#INTRODUCTION)<br>**
- **[OBTAIN](#OBTAIN)**<br>
- **[SCRUB](#SCRUB)**<br>
- **[EXPLORE](#EXPLORE)**<br>
- **[MODEL](#MODEL)**<br>
- **[iNTERPRET](#iNTERPRET)**<br>
- **[Conclusions/Recommendations](#CONCLUSIONS-&-RECOMMENDATIONS)<br>**
___

# INTRODUCTION

> Explain the point of your project and what question you are trying to answer with your modeling.

## Business Problem

Summary of the business problem you are trying to solve, and the data questions that you plan to answer to solve them.

***
Questions to consider:
* What are the business's pain points related to this project?
* How did you pick the data analysis question(s) that you did?
* Why are these questions important from a business perspective?
***



# OBTAIN

## Data Understanding

Describe the data being used for this project.
***
Questions to consider:
* Where did the data come from, and how do they relate to the data analysis questions?
* What do the data represent? Who is in the sample and what variables are included?
* What is the target variable?
* What are the properties of the variables you intend to use?
***


In [9]:
import pandas as pd

In [10]:
#import data into a dataframe
df = pd.read_csv('https://raw.githubusercontent.com/ebtezcan/dsc-phase-3-project/master/data/SpotifyFeatures.csv')
df.head()

genre        artist_name  ... time_signature valence
0  Movie     Henri Salvador  ...            4/4   0.814
1  Movie  Martin & les fées  ...            4/4   0.816
2  Movie    Joseph Williams  ...            5/4   0.368
3  Movie     Henri Salvador  ...            4/4   0.227
4  Movie       Fabien Nataf  ...            4/4   0.390

[5 rows x 18 columns]

In [11]:
df.describe()

popularity   acousticness  ...          tempo        valence
count  232725.000000  232725.000000  ...  232725.000000  232725.000000
mean       41.127502       0.368560  ...     117.666585       0.454917
std        18.189948       0.354768  ...      30.898907       0.260065
min         0.000000       0.000000  ...      30.379000       0.000000
25%        29.000000       0.037600  ...      92.959000       0.237000
50%        43.000000       0.232000  ...     115.778000       0.444000
75%        55.000000       0.722000  ...     139.054000       0.660000
max       100.000000       0.996000  ...     242.903000       1.000000

[8 rows x 11 columns]

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232725 entries, 0 to 232724
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   genre             232725 non-null  object 
 1   artist_name       232725 non-null  object 
 2   track_name        232725 non-null  object 
 3   track_id          232725 non-null  object 
 4   popularity        232725 non-null  int64  
 5   acousticness      232725 non-null  float64
 6   danceability      232725 non-null  float64
 7   duration_ms       232725 non-null  int64  
 8   energy            232725 non-null  float64
 9   instrumentalness  232725 non-null  float64
 10  key               232725 non-null  object 
 11  liveness          232725 non-null  float64
 12  loudness          232725 non-null  float64
 13  mode              232725 non-null  object 
 14  speechiness       232725 non-null  float64
 15  tempo             232725 non-null  float64
 16  time_signature    23

In [13]:
df[df['popularity']>=85]

genre      artist_name  ... time_signature valence
9026    Dance    Ariana Grande  ...            4/4  0.3350
9027    Dance    Ariana Grande  ...            4/4  0.3190
9028    Dance           Halsey  ...            4/4  0.5330
9029    Dance    Ariana Grande  ...            4/4  0.1950
9030    Dance    Ariana Grande  ...            4/4  0.4700
...       ...              ...  ...            ...     ...
166066   Rock  Imagine Dragons  ...            4/4  0.0827
166068   Rock  Imagine Dragons  ...            4/4  0.5560
166147   Rock       Jason Mraz  ...            4/4  0.7180
166850   Rock     ChocQuibTown  ...            4/4  0.8560
191939   Soul      John Legend  ...            4/4  0.3310

[417 rows x 18 columns]

In [14]:
for col in df.columns:
    print(f"Column: {col}")
    print(df[col].value_counts())
    print("--------------------")

Column: genre
Comedy              9681
Soundtrack          9646
Indie               9543
Jazz                9441
Pop                 9386
Electronic          9377
Children’s Music    9353
Folk                9299
Hip-Hop             9295
Rock                9272
Alternative         9263
Classical           9256
Rap                 9232
World               9096
Soul                9089
Blues               9023
R&B                 8992
Anime               8936
Reggaeton           8927
Ska                 8874
Reggae              8771
Dance               8701
Country             8664
Opera               8280
Movie               7806
Children's Music    5403
A Capella            119
Name: genre, dtype: int64
--------------------
Column: artist_name
Giuseppe Verdi            1394
Giacomo Puccini           1137
Kimbo Children's Music     971
Nobuo Uematsu              825
Richard Wagner             804
                          ... 
STEELHEART                   1
Dawn Penn                  

It seems like we have duplicated tracks based on the value counts for the track_id column.

### Changing "Children's Music" characters for consistency

In [15]:
df['genre'].value_counts()

Comedy              9681
Soundtrack          9646
Indie               9543
Jazz                9441
Pop                 9386
Electronic          9377
Children’s Music    9353
Folk                9299
Hip-Hop             9295
Rock                9272
Alternative         9263
Classical           9256
Rap                 9232
World               9096
Soul                9089
Blues               9023
R&B                 8992
Anime               8936
Reggaeton           8927
Ska                 8874
Reggae              8771
Dance               8701
Country             8664
Opera               8280
Movie               7806
Children's Music    5403
A Capella            119
Name: genre, dtype: int64

There are 2 types of "Children's Music" values in the genres due to the character used for apostrophe. Since both Children’s Music and Children's Music is meant to show the same thing we need to change one of them to the other one to be consistent.

In [16]:
df.loc[df['genre']=="Children’s Music",'genre']="Children's Music"

In [17]:
df['genre'].value_counts()

Children's Music    14756
Comedy               9681
Soundtrack           9646
Indie                9543
Jazz                 9441
Pop                  9386
Electronic           9377
Folk                 9299
Hip-Hop              9295
Rock                 9272
Alternative          9263
Classical            9256
Rap                  9232
World                9096
Soul                 9089
Blues                9023
R&B                  8992
Anime                8936
Reggaeton            8927
Ska                  8874
Reggae               8771
Dance                8701
Country              8664
Opera                8280
Movie                7806
A Capella             119
Name: genre, dtype: int64

### Missing Values

In [18]:
df.isna().sum()

genre               0
artist_name         0
track_name          0
track_id            0
popularity          0
acousticness        0
danceability        0
duration_ms         0
energy              0
instrumentalness    0
key                 0
liveness            0
loudness            0
mode                0
speechiness         0
tempo               0
time_signature      0
valence             0
dtype: int64

### Addressing Duplicated Tracks

In [19]:
df[df['track_id'].duplicated()]

genre   artist_name  ... time_signature valence
1348    Alternative      Doja Cat  ...            4/4   0.700
1385    Alternative   Frank Ocean  ...            5/4   0.370
1452    Alternative   Frank Ocean  ...            4/4   0.130
1554    Alternative    Steve Lacy  ...            4/4   0.634
1634    Alternative      tobi lou  ...            4/4   0.599
...             ...           ...  ...            ...     ...
232715         Soul    Emily King  ...            4/4   0.436
232718         Soul  Muddy Waters  ...            4/4   0.945
232720         Soul         Slave  ...            4/4   0.962
232722         Soul  Muddy Waters  ...            4/4   0.813
232723         Soul       R.LUM.R  ...            4/4   0.489

[55951 rows x 18 columns]

In [20]:
len(df[df['track_id'].duplicated()]['track_id'].unique())

35124

We have 55,951 rows of data for 35,124 unique tracks. This means that we have 20,827 rows that we need to address. Before we can address these duplications though we need to see what the cause of the duplicates are.

In [21]:
df[df['track_id']=='6iOvnACn4ChlAw4lWUU4dd'] 

genre artist_name  ... time_signature valence
257                  R&B    Doja Cat  ...            4/4     0.7
1348         Alternative    Doja Cat  ...            4/4     0.7
77710   Children's Music    Doja Cat  ...            4/4     0.7
93651              Indie    Doja Cat  ...            4/4     0.7
113770               Pop    Doja Cat  ...            4/4     0.7

[5 rows x 18 columns]

In [22]:
df[df['track_id']=='2XGLdVl7lGeq8ksM6Al7jT']

genre artist_name    track_name  ...    tempo  time_signature  valence
179212  Jazz       Slave  Son Of Slide  ...  115.542             4/4    0.962
232720  Soul       Slave  Son Of Slide  ...  115.542             4/4    0.962

[2 rows x 18 columns]

In [23]:
df[df['track_id']=='2HFczeynfKGiM9KF2z2K7K']


genre   artist_name  ... time_signature valence
48555   Blues  Muddy Waters  ...            4/4   0.945
232718   Soul  Muddy Waters  ...            4/4   0.945

[2 rows x 18 columns]

In [24]:
df[df['track_id']=='5cA0vB8c9FMOVDWyJHgf26']

genre artist_name track_name  ...   tempo  time_signature  valence
85632    R&B  Emily King       Down  ...  90.831             4/4    0.436
232715  Soul  Emily King       Down  ...  90.831             4/4    0.436

[2 rows x 18 columns]

We see that most of the attributes of the song are the same for the duplicated values except for 'popularity' and 'genre'. The 'popularity' column can be aggregated since it is a numerical column but the categorical column of 'genre' is a little bit trickier. What makes the most sense in this case would be to create different columns with the genre names and display with binary values whether a song belongs to that genre or not.

In [25]:
#generating a list with the genre names
genre_list = list(df['genre'].unique())

In [26]:
#creating the genre columns using the genre list
for genre in genre_list:
    df[genre] = (df['genre']==genre).astype('int')

In [ ]:
#grouping by track_id number to get rid of duplicates and keeping the maximum values in each column.
df=df.groupby(['track_id']).max()

Above, we created the genre columns and merged the duplicated values keeping the maximum value in each column. This makes sense since the track that is being listened to is the same one. If a track's best popularity score was 42 for example, we are keeping the best value by taking the max.

In [ ]:
#removing redundant genre column
df.drop('genre', axis=1, inplace=True)
df.head()

In [ ]:
#verifying that duplicates have been eliminated
df[df.index =='6iOvnACn4ChlAw4lWUU4dd']

In [ ]:
df

We successfully addressed the duplicates and now have 176,774 unique tracks in our dataset (down from 232,725).

## Feature Engineering - is_popular

Since our goal is to be able to identify which tracks will be popular, we need to feature engineer a new column by binarizing the popularity column. To be able to do this, we need to decide on a cut-off point of popularity score which if a song stays above this cut-off point it will be considered "popular" and if it stays below it will be considered "not popular". We can start off by taking a look at the distribution of the popularity score distribution.

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
#creating a histogram to see distribution of popularity scores in the dataset.
sns.histplot(df['popularity'], bins='auto')

From the above histogram we see that we have a bimodal distribution. One of the peaks is at 0, and the other one seems to be around 40. In order to better decide what's popular, we can take a look at the Top 50 songs' popularity scores (this data is also from 2019 similar to our main dataset)

### Top 50 Songs - 2019

In [ ]:
#data from https://www.kaggle.com/leonardopena/top50spotify2019
df_50 = pd.read_csv('https://raw.githubusercontent.com/ebtezcan/dsc-phase-3-project/master/data/top50.csv', encoding='latin1', index_col=0)

In [ ]:
df_50.head()

In [ ]:
#displaying stats information of Top 50 songs
df_50['Popularity'].describe()

Going back to our histogram we can draw vertical lines to see where these values fall into.

In [ ]:
fig, ax = plt.subplots()
sns.histplot(df['popularity'], bins='auto', ax=ax)
stats=['mean', '50%', 'min', 'max']
for stat in stats:
    ax.vlines(x=df_50['Popularity'].describe()[stat], ymin=0, ymax=6000, linestyles='dashed', colors='red', label=stat)

We can see that there was a range of popularity scores in the Top 50 songs between 70 and 95. Which means that any song that is above a 70 theoretically could be a popular song. It doesn't make sense to use median or mean scores for our cutoff point in this case since then we would be disregarding all the songs that had lower values than 87.5 or 88 as unpopular which is untrue. However, before we can establish the cutoff point we need to acknowledge that we are basing it off of only 50 datapoints which is not a lot. It may be good to take a look at Top 100 songs instead of 50 to get a better sample size of popular songs.

### Top 100 Songs - 2019

In [ ]:
#importing dataset
df_100 = pd.read_csv('https://raw.githubusercontent.com/ebtezcan/dsc-phase-3-project/master/data/spotify_top_100_2019.csv')

In [ ]:
df_100['popularity '].describe()

The minimum value of 4 for the popularity score on the Top 100 Songs chart seems like an outlier. Next, we'll visualize the spread of this column to confirm.

In [ ]:
fig, ax = plt.subplots()
sns.histplot(df_100['popularity '], bins='auto', ax=ax)

As we imagined the scores within the range 0-25 seem like outliers. We can remove outliers from this dataset with the IQR method to get a better perspective on the data.

In [ ]:
#Outlier Removal with the IQR method

def find_outliers_IQR(data):
    """Use Tukey's Method of outlier removal AKA InterQuartile-Range Rule
    and return boolean series where True indicates it is an outlier.
    - Calculates the range between the 75% and 25% quartiles
    - Outliers fall outside upper and lower limits, using a treshold of  1.5*IQR the 75% and 25% quartiles.

    IQR Range Calculation:    
        res = df.describe()
        IQR = res['75%'] -  res['25%']
        lower_limit = res['25%'] - 1.5*IQR
        upper_limit = res['75%'] + 1.5*IQR

    Args:
        data (Series,or ndarray): data to test for outliers.

    Returns:
        [boolean Series]: A True/False for each row use to slice outliers.
        
    EXAMPLE USE: 
    >> idx_outs = find_outliers_df(df['AdjustedCompensation'])
    >> good_data = df[~idx_outs].copy()
    
    function snippet from Flatiron School Phase #2 Py Files.
    URL = https://github.com/flatiron-school/Online-DS-FT-022221-Cohort-Notes/blob/master/py_files/functions_SG.py
    
    """
    df_b=data
    res= df_b.describe()

    IQR = res['75%'] -  res['25%']
    lower_limit = res['25%'] - 1.5*IQR
    upper_limit = res['75%'] + 1.5*IQR

    idx_outs = (df_b>upper_limit) | (df_b<lower_limit)

    return idx_outs

In [ ]:
#removing outliers from the popularity column
df_100 = df_100[find_outliers_IQR(df_100['popularity '])==False]
#displaying minimum & maxium values in popularity column 
print("Minimum:", df_100['popularity '].min())
print("Maximum:", df_100['popularity '].max())

In [ ]:
#visualizing the min and max popularity scores on the overall dataset histogram
fig, ax = plt.subplots()
sns.histplot(df['popularity'], bins='auto', ax=ax)
ax.vlines(x=df_100['popularity '].min(), ymin=0, ymax=6000, linestyles='dashed', colors='red', label='min')
ax.vlines(x=df_100['popularity '].max(), ymin=0, ymax=6000, linestyles='dashed', colors='red', label='max')
plt.legend()

As we can expect to see, the top 100 songs have a lower popularity score threshold compared to the top 50 songs. We will be defining a song being popular as being Top 100 worthy and therefore will establish our cutoff point at 58.

In [ ]:
#creating is_popular column with our cutoff point
df['is_popular']=(df['popularity']>=58).astype('int')
df.head()

In [ ]:
#dropping popularity score column since we will not be using it
df.drop(['popularity', 'artist_name', 'track_name'], axis=1, inplace=True)
df.head()

We dropped popularity scores since we already binarized that column, but additionally we are dropping 'artist_name' and 'track_name' since we are looking at the anatomy of a song and not who sings it or what it's called.

## One Hot Encoding the Categorical Columns

We still have categorical columns that need one hot encoding. Namely, these columns are 'key', 'mode', 'time_signature'.

In [ ]:
#Check to see how many more columns we will be creating by OHE the cat_cols.
df.nunique()

We will be creating 2 (mode) + 5 (time_signature) + key (12) - 3 (drop_first) = 16 columns.

In [ ]:
cat_cols = ['key', 'mode', 'time_signature']

In [ ]:
from sklearn.preprocessing import OneHotEncoder

encoder=OneHotEncoder(sparse=False, drop='first')
data_ohe = encoder.fit_transform(df[cat_cols])
df_ohe = pd.DataFrame(data_ohe, columns=encoder.get_feature_names(cat_cols), index=df.index)

In [ ]:
pd.set_option("display.max_columns", None)
df_ohe

In [ ]:
df_ohe = pd.concat([df.drop(cat_cols, axis=1), df_ohe], axis=1)
df_ohe.head()

## train_test_split

In [ ]:
from sklearn.model_selection import train_test_split
y=df_ohe['is_popular']
X=df_ohe.drop('is_popular',axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, )

## Model #1 - Baseline - Dummy Classifier

In [ ]:
from sklearn.dummy import DummyClassifier

clf_dummy = DummyClassifier()
clf_dummy.fit(X_train, y_train)
y_pred = clf_dummy.predict(X_test)

We need a function that will show us the classification report, the confusion matrix as well as the ROC curve to be able to evaluate our models.

In [ ]:
from sklearn.metrics import classification_report, plot_confusion_matrix, plot_roc_curve

def classification(y_true, y_pred, X, clf):
    """This function shows the classification report,
    the confusion matrix as well as the ROC curve for evaluation of model quality.
    
    y_true: Correct y values, typically y_test that comes from the train_test_split performed at the beginning of model development.
    y_pred: Predicted y values by the model.
    clf: classifier model that was fit to training data.
    X: X_test values"""
    
    #Classification report
    print("CLASSIFICATION REPORT")
    print("------------------------------------------")
    print(classification_report(y_true=y_true, y_pred=y_pred))
    
    #Creating a figure/axes for confusion matrix and ROC curve
    fig, ax = plt.subplots(ncols=2, figsize=(12, 5))
    
    #Plotting the normalized confusion matrix
    plot_confusion_matrix(estimator=clf, X=X, y_true=y_true, cmap='Blues', normalize='true', ax=ax[0])
    
    #Plotting the ROC curve
    plot_roc_curve(estimator=clf, X=X, y=y_true, ax=ax[1])
    
    #Plotting the 50-50 guessing plot for reference
    ax[1].plot([0,1], [0,1], ls='--', color='orange')

In [ ]:
classification(y_test, y_pred, X_test, clf_dummy)

In [ ]:
#class imbalance percentages
y_train.value_counts(normalize=True)

Our dummy classifier correctly predicted 98% of the unpopular songs as unpopular; however, it failed to correctly predict the popular songs as popular and instead classified them as unpopular as well. We clearly have a class imbalance problem where approximately 98% of our data is not popular and only about 2% of it is. To address this we can SMOTE the training data and see if training a model with this method would improve our results.

## Addressing Class Imbalance with SMOTENC

In [ ]:
#looking at column names to extract categorical column indices for SMOTENC
X.columns

In [ ]:
#creating a list of categorical column indices
cat_cols = list(range(10, len(X.columns)))
X.columns[cat_cols]

In [ ]:
from imblearn.over_sampling import SMOTE, SMOTENC

sm = SMOTENC(categorical_features=cat_cols)

X_train_sm, y_train_sm = sm.fit_resample(X_train, y_train)
y_train_sm.value_counts(normalize=True)

## Model #2 - Random Forest Classifier

### Initial Model

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

clf = RandomForestClassifier()
clf.fit(X_train_sm, y_train_sm)

y_pred = clf.predict(X_test)
classification(y_test, y_pred, X_test, clf)

In [ ]:
#Evaluating the model performance for the training data
clf = RandomForestClassifier()
clf.fit(X_train_sm, y_train_sm)

y_pred = clf.predict(X_train_sm)
classification(y_train_sm, y_pred, X_train_sm, clf)

Our model is performing very well on the training data but not so much on the test data since it is overfitting to the training set. We need to tune our model to get more accurate results on unseen data.

### Hyperparameter Tuning

In [ ]:
np.linspace(start = 200, stop = 2000, num = 10)

In [ ]:
import numpy as np
from sklearn.model_selection import RandomizedSearchCV

clf = RandomForestClassifier()
max_depth = [int(x) for x in np.linspace(10, 110, num = 11)]
n_estimators = [int(x) for x in np.linspace(start = 200, stop = 2000, num = 10)]
param_dist = {'max_depth': max_depth,
              'bootstrap': [True, False]}

random_search = RandomizedSearchCV(estimator=clf, param_distributions = param_dist, scoring='f1', n_jobs=3, cv=3)
random_search.fit(X_train_sm, y_train_sm)
random_search.best_params_

In [ ]:
from sklearn.model_selection import GridSearchCV

clf = RandomForestClassifier()
grid = {'max_leaf_nodes': [100, 300, 600], 
        'max_depth': [1, 10, 100],
       }

gridsearch = GridSearchCV(estimator=clf, param_grid = grid, scoring='f1')

gridsearch.fit(X_train_sm,  y_train_sm)
gridsearch.best_params_

In [ ]:
clf = RandomForestClassifier(max_leaf_nodes=600, max_depth=100)
clf.fit(X_train_sm, y_train_sm)

y_pred = clf.predict(X_test)
classification(y_test, y_pred, X_test, clf)

## Model #3 - LogRegCV

### Removing Outliers

In [ ]:
#separating out the numerical columns for outlier removal
num_cols = X.columns[0:10]
num_cols

In [ ]:
for col in num_cols:
    df_ohe = df_ohe[find_outliers_IQR(df_ohe[col])==False]
    
df_ohe

### train_test_split 

In [ ]:
y=df_ohe['is_popular']
X=df_ohe.drop('is_popular', axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

### Addressing Class Imbalance with SMOTE

In [ ]:
y_train.value_counts(normalize=True)

In [ ]:
sm = SMOTENC(categorical_features=cat_cols)

X_train_sm, y_train_sm = sm.fit_resample(X_train, y_train)
y_train_sm.value_counts(normalize=True)

### Scaling the Data

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_sm_sc = scaler.fit_transform(X_train_sm)
X_test_sc = scaler.transform(X_test)

### Initial Model

In [ ]:
from sklearn.linear_model import LogisticRegressionCV, LogisticRegression

We will start by exploring how a default logistic regression model performs on our data. The only parameter we are changing is class_weight since we used SMOTENC on our data and we know that we have equal representation across the board.

In [ ]:
clf = LogisticRegression(class_weight='balanced')
clf.fit(X_train_sm_sc, y_train_sm)
y_pred = clf.predict(X_test_sc)
classification(y_test, y_pred, X_test_sc, clf)

It seems like our model may be overfitting to the training data. We can take a look at how the model performs on the training data to verify.

In [ ]:
#Evaluating the model performance for the training data
y_pred = clf.predict(X_train_sm_sc)
classification(y_train_sm, y_pred, X_train_sm_sc, clf)

As we guessed our model is overfitting to the training data and performs very well on it while the performance on the test data is subpar. In order to address this issue we can move onto hyperparameter tuning with GridSearchCV. We will be trying to optimize our grid search for the f1 score since optimizing for f1 score for our dataset will find the right balance between precision and recall and be the best overall performance score for the model.

### Hyperparameter Tuning

In [ ]:
from sklearn.model_selection import GridSearchCV

clf = LogisticRegression()
grid = {'C': [1e12, 1000, 100, 1], 
        'penalty': ['l1','l2'],
        'solver': ['liblinear', 'lbfgs', 'sag', 'saga']
       }

gridsearch = GridSearchCV(estimator=clf, param_grid = grid, scoring='f1')

gridsearch.fit(X_train_sm_sc,  y_train_sm)
gridsearch.best_params_

In [ ]:
clf = LogisticRegression(class_weight='balanced')
grid = {'C': [1e12, 1000, 100, 10, 1],
        'penalty': ['l1', 'l2'],
        'solver': ['lbfgs', 'saga']
       }

gridsearch = GridSearchCV(estimator=clf, param_grid = grid, scoring='f1')

gridsearch.fit(X_train_sm_sc,  y_train_sm)
gridsearch.best_params_

### Tuned LogisticRegression model

In [ ]:
clf = LogisticRegression(class_weight='balanced', C=1e12, penalty='l1', solver='saga', max_iter=1000)
clf.fit(X_train_sm_sc, y_train_sm)
y_pred = clf.predict(X_test_sc)
classification(y_test, y_pred, X_test_sc, clf)

In [ ]:
clf = LogisticRegression(class_weight='balanced', C=1e12, penalty='l2', solver='lbfgs', max_iter=100)
clf.fit(X_train_sm_sc, y_train_sm)
y_pred = clf.predict(X_test_sc)
classification(y_test, y_pred, X_test_sc, clf)

We still have the issue of overfitting. We can try out a LogisticRegressionCV model to cross validate the model and see if this approach improves our results. We'll start with a basic model with only the cv and class_weight parameters provided.

## LogisticRegressionCV

In [ ]:
from sklearn.linear_model import LogisticRegressionCV
clf = LogisticRegressionCV(cv=5, class_weight='balanced')
clf.fit(X_train_sm_sc,  y_train_sm)
y_pred = clf.predict(X_test_sc)
classification(y_test, y_pred, X_test_sc, clf)

In [ ]:
#Evaluating the model performance for the training data
clf = LogisticRegressionCV(cv=5, class_weight='balanced')
clf.fit(X_train_sm_sc,  y_train_sm)
y_pred = clf.predict(X_train_sm_sc)
classification(y_train_sm, y_pred, X_train_sm_sc, clf)

### Hyperparameter Tuning - LogisticRegressionCV

In [ ]:
clf = LogisticRegressionCV(class_weight='balanced', max_iter=1000)
grid = {'penalty': ['l1','l2'],
        'solver': ['lbfgs', 'saga']
   }

gridsearch = GridSearchCV(estimator=clf, param_grid = grid, scoring='f1')

gridsearch.fit(X_train_sm_sc,  y_train_sm)
gridsearch.best_params_

## XGBoost

In [ ]:
from xgboost import XGBClassifier

In [ ]:
clf = XGBClassifier()
clf.fit(X_train_sm, y_train_sm)
y_pred = clf.predict(X_test)
classification(y_test, y_pred, X_test, clf)

In [ ]:
grid = {
    'learning_rate': [0.1, 0.2],
        }
gridsearch = GridSearchCV(estimator=clf, param_grid = grid, scoring='f1')

gridsearch.fit(X_train_sm,  y_train_sm)
gridsearch.best_params_

# SCRUB

## Data Preparation

Describe and justify the process for preparing the data for analysis.

***
Questions to consider:
* Were there variables you dropped or created?
* How did you address missing values or outliers?
* Why are these choices appropriate given the data and the business problem?
***

# EXPLORE

# MODEL

## Data Modeling
Describe and justify the process for analyzing or modeling the data.

***
Questions to consider:
* How did you analyze or model the data?
* How did you iterate on your initial approach to make it better?
* Why are these choices appropriate given the data and the business problem?
***

# iNTERPRET

Evaluate how well your work solves the stated business problem.

***
Questions to consider:
* How do you interpret the results?
* How well does your model fit your data? How much better is this than your baseline model?
* How confident are you that your results would generalize beyond the data you have?
* How confident are you that this model would benefit the business if put into use?
***

# CONCLUSIONS & RECOMMENDATIONS

Provide your conclusions about the work you've done, including any limitations or next steps.

***
Questions to consider:
* What would you recommend the business do as a result of this work?
* What are some reasons why your analysis might not fully solve the business problem?
* What else could you do in the future to improve this project?
***